src/sdam/server.ts

```diff
@@ -338,46 +317,122 @@ export class Server extends TypedEventEmitter<ServerEvents> {
-      this.pool.checkOut((err, checkedOut) => {
-        if (err || checkedOut == null) {
-          if (callback) return callback(err);
+      const { promise: checkedOutPromise, resolve, reject } = promiseWithResolvers<Connection>();
+
+      this.pool.checkOut((err, conn) => {
+        if (err || conn == null) {
+          reject(err);
           return;
         }
-
-        session.pin(checkedOut);
-        this.command(ns, cmd, finalOptions, callback);
+        resolve(conn);
       });
-      return;
+      const checkedOut = await checkedOutPromise;
+
+      session.pin(checkedOut);
+      return this.command(ns, cmd, finalOptions);
     }
 
     this.incrementOperationCount();
 
-    this.pool.withConnection(
-      conn,
-      (err, conn, cb) => {
-        if (err || !conn) {
-          this.decrementOperationCount();
-          if (!err) {
-            return cb(new MongoRuntimeError('Failed to create connection without error'));
-          }
-          if (!(err instanceof PoolClearedError)) {
-            this.handleError(err);
-          }
-          return cb(err);
+    /*
+    const executeCommand: WithConnectionCallBack = (err, conn, cb) => {
+      if (err || !conn) {
+        this.decrementOperationCount();
+        if (!err) {
+          return cb(new MongoRuntimeError('Failed to create connection without error'));
         }
+        if (!(err instanceof PoolClearedError)) {
+          this.handleError(err);
+        }
+        return cb(err);
+      }
 
-        const handler = makeOperationHandler(this, conn, cmd, finalOptions, (error, response) => {
-          this.decrementOperationCount();
-          cb(error, response);
-        });
-        conn.command(ns, cmd, finalOptions).then(
-          r => handler(undefined, r),
-          e => handler(e)
-        );
-      },
-      callback
-    );
+      const handler = makeOperationHandler(this, conn, cmd, finalOptions, (error, response) => {
+        this.decrementOperationCount();
+        cb(error, response);
+      });
+      conn.command(ns, cmd, finalOptions).then(
+        r => handler(undefined, r),
+        e => handler(e)
+      );
+    };
+    */
+
+    const executeCommandAsync = async (err?: AnyError, conn?: Connection) => {
+      if (err || !conn) {
+        this.decrementOperationCount();
+        if (!err) throw new MongoRuntimeError('Failed to create connection without error');
+        if (!(err instanceof PoolClearedError)) this.handleError(err);
+
+        throw err;
+      }
+      const handler = makeAsyncOperationHandler<Document>(this, conn, cmd, finalOptions);
+
+      try {
+        const result = await conn.command(ns, cmd, finalOptions);
+        return await handler(undefined, result);
+      } catch (e) {
+        return await handler(e);
+      }
+    };
+    const shouldReauth = (e: AnyError): boolean => {
+      return e instanceof MongoError && e.code === MONGODB_ERROR_CODES.Reauthenticate;
+    };
```

Comment:
NOTE: This is the asyncified version of the callback previously passed to `this.pool.checkOut`